src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -751,114 +741,219 @@ private void ReturnReadBufferIfEmpty()
                 _decryptedBytesCount = 0;
                 _decryptedBytesOffset = 0;
             }
+            else if (_decryptedBytesCount == 0)
+            {
+                _decryptedBytesOffset = 0;
+            }
         }
 
-        private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, Memory<byte> buffer)
+        private bool HaveFullTlsFrame(out int frameSize)
+        {
+            if (_internalBufferCount < SecureChannel.ReadHeaderSize)
+            {
+                frameSize = int.MaxValue;
+                return false;
+            }
+
+            frameSize = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
+            return _internalBufferCount >= frameSize;
+        }
+
+        private ValueTask<int> GetFullFrameIfNeed<TIOAdapter>(TIOAdapter adapter)
```

Comment:
In other places we use the term "Ensure" here. Can we call this something like "EnsureFullFrame"?

Also, it should be consistent with HaveFullTlsFrame. So either "Frame" or "TlsFrame" for both.

Reply:
Updated to EnsureFullTlsFrame